### PR TITLE
chore: add xray report generation capability

### PR DESCRIPTION
**Pull request by blaisedias**

### CODE REVIEW

File: scripts/xray-report.py [hunk 1]
```diff
@@ -0,0 +1,616 @@
...
+        self.txt = txt
+        self.link = link
+        if titl is not None:
+            titl.replace('"', '&quot;')
+            titl.replace('<', '&lt;')
+            titl.replace('>', '&gt;')
+            titl = html.escape(titl)
+        self.titl = titl
+        self.colspan = colspan
+        self._ch = 'td'
```
Comment (chriswldenyer) [inline scripts/xray-report.py:118]: What does ch stand for?

File: scripts/xray-report.py [hunk 2]
```diff
@@ -0,0 +1,616 @@
...
+
+
+def orderTests(tests):
+    kiis = sorted(
+        tests.keys(),
+        key=lambda test: int(test.split('-')[-1])
+    )
+
+    ordered = []
+    for k in kiis:
```
Comment (chriswldenyer) [inline scripts/xray-report.py:229]: This could do with a few comments

File: scripts/xray-report.py [hunk 3]
```diff
@@ -0,0 +1,616 @@
...
+                    txt=test['status']['name'][0],
+                    titl=f'{xtime}\n\n{test["results"][0]["log"]}',
+                    link=runLink(exe["jira"]["key"], testKey),
+                )
+            row += 1
+        col += 1
+
+    gradeback = 14
+    scores = {}
+    for ix in [x for x in range(len(atable))
```
Comment (chriswldenyer) [inline scripts/xray-report.py:420]: Should mention that this is ordering rows based on the frequency of failures.

File: scripts/xray-report.py [hunk 4]
```diff
@@ -0,0 +1,616 @@
...
+        [headerCell(txt="Failed Test Count")],
+        [headerCell(txt="Total")],
+        [headerCell(txt="Not run Test Count")],
+        [headerCell(txt="Failed Before/After Suite")],
+        [headerCell(txt="Not run Before/After Suite")],
+    ]
+    for iex in range(len(executions)):
+        # offset for header column
+        ix = iex + 1
+        fc = 0
```
Comment (chriswldenyer) [inline scripts/xray-report.py:472]: These could do with better names, I'll guess failure count and pass count, not sure about the rest.

**Review by chriswldenyer — APPROVED** [review]
Looks plausible. A few comments.

File: scripts/xray-report.py (hunk 1 above)
Comment (blaisedias) [reply scripts/xray-report.py:118]: I have changed from cryptic `_ch` to `celltype`- it ie either `<td>` or `<th>`

File: scripts/xray-report.py (hunk 2 above)
Comment (blaisedias) [reply scripts/xray-report.py:229]: Done.

File: scripts/xray-report.py [hunk 5]
```diff
@@ -0,0 +1,616 @@
...
+    # Unfortunately there isn't a field we can use to derive the
+    # timestamp for the test execution. So we record the earliest
+    # timestamp of a test run in the test execution, and use that
+    # to generate the date information for the execution
+    for exe in executions:
+        execDict = {
+            d['test']['jira']['key']: d for d in jsonLoad(
+                f'{_args.cachedir}/exec.{exe["jira"]["key"]}.json')
+        }
+        ts = 0
```
[the remark this answers is not in the document]
Comment (blaisedias) [reply scripts/xray-report.py:317]: renamed to timestamp.

File: scripts/xray-report.py (hunk 4 above)
Comment (blaisedias) [reply scripts/xray-report.py:472]: Done is now
```
        fail_count = 0
        pass_count = 0
        notrun_count = 0
        beforeaftersuite_notrun_count = 0
        beforeaftersuite_fail_count = 0
```

File: scripts/analyse.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+re_It = re.compile(
+    r'^(\s*|\s*ginkgo\.)It\s*\(\s*"(?P<it>.*)"\s*,\s*func\s*\(\s*\)\s*{')
+re_ItPromiscuous = re.compile(
+    r'^(\s*|\s*ginkgo\.)It\s*\(\s*(?P<it>.*),\s*func\s*\(\s*\)\s*{')
+re_RunSpecs = re.compile(
+    r'^\s*ginkgo.RunSpecsWithDefaultAndCustomReporters\s*\(\s*\w\s*,\s*"(?P<classname>.*)"\s*,')
+
+
+def scrapeSourceFileClassnames(f, classnames, verbose=False):
+    """ it should be obvious from the source
```
[the remark this answers is not in the document]
Comment (blaisedias) [reply scripts/analyse.py]: Done.

File: scripts/analyse.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
+#!/usr/bin/env python
+import re
+import sys
+import os
+
+re_InitTesting = re.compile(
+    r'^\s*k8stest.InitTesting\((?P<t>.*)\s*,\s*"(?P<classname>.*)"\s*,\s*(?P<reportname>.*)\s*\)')
```
[the remark this answers is not in the document]
Comment (blaisedias) [reply scripts/analyse.py:9]: Done.

File: scripts/xray-report.py (hunk 3 above)
Comment (blaisedias) [reply scripts/xray-report.py:437]: Done.

**Pull request merged**